### PR TITLE
Add sorting to session and cohort views

### PR DESCRIPTION
**Pull request by benilovj**
* sessions with dates are sorted by earliest date and location name
* sessions without dates are sorted by location name
* cohorts are sorted by patient first and last names

### CODE REVIEW

File: app/controllers/programmes_controller.rb [hunk 1]
```diff
@@ -14,9 +14,16 @@ def show
...
+    sessions_for_programme =
+      policy_scope(Session)
+        .has_programme(@programme)
+        .strict_loading
+        .includes(:dates, :location)
 
-    @scheduled_sessions = sessions_for_programme.scheduled
+    @scheduled_sessions =
+      sessions_for_programme.scheduled.sort_by do |session|
+        [session.dates.first.value, session.location.name]
```
Comment (thomasleese) [inline app/controllers/programmes_controller.rb:25]: The location can be `nil` if we import a session for an unknown school so I think we'll want to handle that here.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -21,31 +21,55 @@ def new
   end
 
   def index
-    @sessions = policy_scope(Session).today
+    @sessions =
+      policy_scope(Session)
+        .strict_loading
+        .includes(:dates, :location, :programmes)
+        .today
+        .sort_by { |session| session.location.name }
```
Comment (thomasleese) [inline app/controllers/sessions_controller.rb:29]: Could we do the sorting at the database level to improve performance or maybe even have a default ordering?

Comment (benilovj) [reply app/controllers/sessions_controller.rb:29]: Do you mean something like this?

```
diff --git a/app/controllers/sessions_controller.rb b/app/controllers/sessions_controller.rb
index ba43c7d3..2249e100 100644
--- a/app/controllers/sessions_controller.rb
+++ b/app/controllers/sessions_controller.rb
@@ -26,7 +26,6 @@ class SessionsController < ApplicationController
         .strict_loading
         .includes(:dates, :location, :programmes)
         .today
-        .sort_by { |session| session.location.name }

     render layout: "full"
   end
diff --git a/app/models/session.rb b/app/models/session.rb
index eec1212a..e0422f28 100644
--- a/app/models/session.rb
+++ b/app/models/session.rb
@@ -50,7 +50,12 @@ class Session < ApplicationRecord
   scope :has_programme,
         ->(programme) { joins(:programmes).where(programmes: programme) }

-  scope :today, -> { has_date(Date.current) }
+  scope :today,
+        -> do
+          has_date(Date.current).left_joins(:location).order(
+            "locations.name ASC NULLS LAST"
+          )
+        end
```

Comment (thomasleese) [reply app/controllers/sessions_controller.rb:29]: Yeah something like that looks good to me!

File: app/controllers/programmes_controller.rb (hunk 1 above)
Comment (benilovj) [reply app/controllers/programmes_controller.rb]: Updated in https://github.com/nhsuk/manage-vaccinations-in-schools/pull/1945/commits/5806f935e4bc7c9a04b2de0fbb18a9f45074909f to handle `location = nil`

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (benilovj) [reply app/controllers/sessions_controller.rb]: 👍 Addressed in https://github.com/nhsuk/manage-vaccinations-in-schools/pull/1945/commits/ee4ef13075fc6c3d08629b3b0e5a5868a4d7baec

**Review by tvararu — APPROVED** [review]

File: app/controllers/programmes_controller.rb [hunk 3]
```diff
@@ -27,8 +34,12 @@ def sessions
           .map do |location|
             Session.new(team: current_user.team, location:, academic_year:)
           end
+    @unscheduled_sessions.sort_by! { |session| session.location.name }
```
Comment (thomasleese) [inline app/controllers/programmes_controller.rb:37]:
```suggestion
    @unscheduled_sessions.sort_by! { |session| session.location&.name }
```

File: app/controllers/programmes_controller.rb [hunk 4]
```diff
@@ -27,8 +34,12 @@ def sessions
...
 
-    @completed_sessions = sessions_for_programme.completed
+    @completed_sessions =
+      sessions_for_programme.completed.sort_by do |session|
+        [session.dates.first.value, session.location.name]
```
Comment (thomasleese) [inline app/controllers/programmes_controller.rb:41]:
```suggestion
        [session.dates.first.value, session.location&.name]
```

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -21,31 +21,54 @@ def new
...
 
   def scheduled
-    @sessions = policy_scope(Session).scheduled
+    @sessions =
+      policy_scope(Session)
+        .strict_loading
+        .includes(:dates, :location, :programmes)
+        .scheduled
+        .sort_by do |session|
+          [session.dates.first.value, session.location.name]
```
Comment (thomasleese) [inline app/controllers/sessions_controller.rb:40]:
```suggestion
          [session.dates.first.value, session.location&.name]
```

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -21,31 +21,54 @@ def new
...
 
   def completed
-    @sessions = policy_scope(Session).completed
+    @sessions =
+      policy_scope(Session)
+        .strict_loading
+        .includes(:dates, :location, :programmes)
+        .completed
+        .sort_by do |session|
+          [session.dates.first.value, session.location.name]
```
Comment (thomasleese) [inline app/controllers/sessions_controller.rb:70]:
```suggestion
          [session.dates.first.value, session.location&.name]
```

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -21,31 +21,54 @@ def new
...
+        .strict_loading
+        .includes(:dates, :location, :programmes)
+        .unscheduled +
         team
           .schools
           .for_year_groups(team.year_groups)
           .has_no_session
           .map { |location| Session.new(team:, location:, academic_year:) }
 
+    @sessions.sort_by! { |session| session.location.name }
```
Comment (thomasleese) [inline app/controllers/sessions_controller.rb:58]:
```suggestion
    @sessions.sort_by! { |session| session.location&.name }
```

File: app/controllers/programmes_controller.rb (hunk 3 above)
Comment (benilovj) [reply app/controllers/programmes_controller.rb]: Ugh, apologies for leaving the inconsistencies in.

Should be fixed now everywhere now.

**Review by thomasleese — APPROVED** [review]

File: app/controllers/programmes_controller.rb (hunk 4 above)
Comment (benilovj) [reply app/controllers/programmes_controller.rb]: Fixed now. I debated whether to also push this sort into the database but held off for now for expediency. We can probably sweet it up when we revisit this code or if we discover it's not performant.

Comment (thomasleese) [reply app/controllers/programmes_controller.rb]: Yeah, that makes sense!

**Pull request merged**